File: Data Structures & Algorithms/combination-target-sum/submission-0.rs
impl Solution {
    pub fn combination_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut res = Vec::new();
        let mut curr = Vec::new();

        // Use a closure to capture 'res', 'nums', and 'target' from the environment
        fn dfs(
            i: usize, 
            total: i32, 
            target: i32, 
            nums: &Vec<i32>, 
            curr: &mut Vec<i32>, 
            res: &mut Vec<Vec<i32>>
        ) {
            if total == target {
                res.push(curr.clone());
                return;
            }

            // Indexing and length checks must use usize
            if i >= nums.len() || total > target {
                return;
            }

            curr.push(nums[i]);
            dfs(i, total + nums[i], target, nums, curr, res);
            curr.pop();
            dfs(i + 1, total, target, nums, curr, res);
        }

        dfs(0, 0, target, &nums, &mut curr, &mut res);
        res
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res: List[List[int]] = []

        nums.sort()

        def dfs(i: int, curr: List[int]):
            if i == len(nums):
                if curr not in res:
                    res.append(curr.copy())
                return
            curr.append(nums[i])
            dfs(i + 1, curr)
            curr.pop()
            dfs(i + 1, curr)
        
        dfs(0, [])
        return res